cmake/grpc_generate_cpp.cmake

```diff
@@ -0,0 +1,104 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# A helper function to generate C++ code from given .proto files.
+cmake_minimum_required(VERSION 3.8)
+find_package(Threads REQUIRED)
+
+function(grpc_generate_cpp SRCS HDRS INCLUDE_DIRS)
+    # Expect:
+    # - PROTOC_EXECUTABLE: path to protoc
+    # - GRPC_CPP_EXECUTABLE: path to grpc_cpp_plugin
+    if(NOT ARGN)
+        message(SEND_ERROR "Error: grpc_generate_cpp() called without any .proto files")
+        return()
+    endif()
+
+    foreach(PROTO_FILE ${ARGN})
+        message(STATUS "Build proto file ${PROTO_FILE}")
+        # Get the full path to the proto file
+        get_filename_component(_abs_proto_file "${PROTO_FILE}" ABSOLUTE)
+        # Get the name of the proto file without extension
+        get_filename_component(_proto_name_we ${PROTO_FILE} NAME_WE)
+        # Get the parent directory of the proto file
+        get_filename_component(_proto_parent_dir ${_abs_proto_file} DIRECTORY)
+        # Get the parent directory of the parent directory
+        get_filename_component(_parent_dir ${_proto_parent_dir} DIRECTORY)
+        # Append 'generated' to the parent directory
+        set(_generated_dir "${_parent_dir}/generated")
+        file(MAKE_DIRECTORY ${_generated_dir})
+
+        set(_protobuf_include_path -I ${_proto_parent_dir})
+
+        set(_proto_srcs "${_generated_dir}/${_proto_name_we}.pb.cc")
+        set(_proto_hdrs "${_generated_dir}/${_proto_name_we}.pb.h")
+        set(_grpc_srcs "${_generated_dir}/${_proto_name_we}.grpc.pb.cc")
+        set(_grpc_hdrs "${_generated_dir}/${_proto_name_we}.grpc.pb.h")
+
+        file(REMOVE "${_proto_srcs}" "${_proto_hdrs}" "${_grpc_srcs}" "${_grpc_hdrs}")
+        add_custom_command(
+            OUTPUT "${_proto_srcs}" "${_proto_hdrs}" "${_grpc_srcs}" "${_grpc_hdrs}"
+            COMMAND ${PROTOC_EXECUTABLE}
+            ARGS --grpc_out=${_generated_dir}
+            --cpp_out=${_generated_dir}
+            --plugin=protoc-gen-grpc=${GRPC_CPP_EXECUTABLE}
+            ${_protobuf_include_path} ${_abs_proto_file}
+            DEPENDS ${_abs_proto_file}
+            COMMENT "Running gRPC C++ protocol buffer compiler on ${PROTO_FILE}"
+            VERBATIM
+        )
+
+        list(APPEND ${SRCS} "${_proto_srcs}")
+        list(APPEND ${HDRS} "${_proto_hdrs}")
+        list(APPEND ${SRCS} "${_grpc_srcs}")
+        list(APPEND ${HDRS} "${_grpc_hdrs}")
+        list(APPEND ${INCLUDE_DIRS} "${_generated_dir}")
+    endforeach()
+
+    set_source_files_properties(${${SRCS}} ${${HDRS}} PROPERTIES GENERATED TRUE)
+    set(${SRCS} "${${SRCS}}" PARENT_SCOPE)
+    set(${HDRS} "${${HDRS}}" PARENT_SCOPE)
+    set(${INCLUDE_DIRS} "${${INCLUDE_DIRS}}" PARENT_SCOPE)
+endfunction()
+
+include(FetchContent)
+set(ABSL_ENABLE_INSTALL ON)
+FetchContent_Declare(
```

Comment:
As an enhancement, it could be nice to inline this block in a `macro` to allow `grpc_generate_cpp.cmake` to be included without automatically adding gRPC to the build, such as if another project wants to use a specific version of gRPC. Not a blocker if you'd rather move forward as-is.